File: app/src/main/java/com/techyourchance/multithreading/demonstrations/designthreadpool/MyBlockingQueue.java
package com.techyourchance.multithreading.demonstrations.designthreadpool;

import java.util.LinkedList;
import java.util.Queue;

/**
 * Simplified implementation of blocking queue.
 */
class MyBlockingQueue {

    private final Object QUEUE_LOCK = new Object();

    private final int mCapacity;
    private final Queue<Integer> mQueue = new LinkedList<>();

    private int mCurrentSize = 0;

    MyBlockingQueue(int capacity) {
        mCapacity = capacity;
    }

    /**
     * Inserts the specified element into this queue, waiting if necessary
     * for space to become available.
     *
     * @param number the element to add
     */
    public void put(int number) {
        synchronized (QUEUE_LOCK) {
            while (mCurrentSize >= mCapacity) {
                try {
                    QUEUE_LOCK.wait();
                } catch (InterruptedException e) {
                    return;
                }
            }

            mQueue.offer(number);
            mCurrentSize++;
            QUEUE_LOCK.notifyAll();
        }
    }

    /**
     * Retrieves and removes the head of this queue, waiting if necessary
     * until an element becomes available.
     *
     * @return the head of this queue
     */
    public int take() {
        synchronized (QUEUE_LOCK) {
            while (mCurrentSize <= 0) {
                try {
                    QUEUE_LOCK.wait();
                } catch (InterruptedException e) {
                    return 0;
                }
            }
            mCurrentSize--;
            QUEUE_LOCK.notifyAll();
            return mQueue.poll();
        }
    }
}

File: app/src/main/java/com/techyourchance/multithreading/demonstrations/designcoroutines/MyBlockingQueue.kt
package com.techyourchance.multithreading.demonstrations.designcoroutines

import java.util.LinkedList
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Simplified implementation of blocking queue.
 */
internal class MyBlockingQueue(private val capacity: Int) {

    private val reentrantLock = ReentrantLock()
    private val lockCondition = reentrantLock.newCondition()

    private val queue = LinkedList<Int>()

    private var currentSize = 0

    /**
     * Inserts the specified element into this queue, waiting if necessary
     * for space to become available.
     *
     * @param number the element to add
     */
    fun put(number: Int) {
        reentrantLock.withLock {
            while (currentSize >= capacity) {
                try {
                    lockCondition.await()
                } catch (e: InterruptedException) {
                    return
                }
            }
            queue.offer(number)
            currentSize++
            lockCondition.signalAll()
        }
    }

    /**
     * Retrieves and removes the head of this queue, waiting if necessary
     * until an element becomes available.
     *
     * @return the head of this queue
     */
    fun take(): Int {
        reentrantLock.withLock {
            while (currentSize <= 0) {
                try {
                    lockCondition.await()
                } catch (e: InterruptedException) {
                    return 0
                }
            }
            currentSize--
            lockCondition.signalAll()
            return queue.poll()
        }
    }
}
